File: src/main/java/burrows/apps/example/gif/di/component/AppComponent.java
package burrows.apps.example.gif.di.component;

import android.app.Application;
import android.content.Context;
import burrows.apps.example.gif.App;
import burrows.apps.example.gif.di.module.AppModule;
import burrows.apps.example.gif.di.module.LeakCanaryModule;
import com.squareup.leakcanary.RefWatcher;
import dagger.Component;

import javax.inject.Singleton;

/**
 * @author <a href="mailto:jaredsburrows@gmail.com">Jared Burrows</a>
 */
@Singleton
@Component(modules = {AppModule.class, LeakCanaryModule.class})
public interface AppComponent {
  // Injections
  void inject(App app);

  // Expose to subgraphs
  Application application();
  Context context();
  RefWatcher refWatcher();

  // Setup components dependencies and modules
  final class Builder {
    private Builder() {
    }

    public static AppComponent build(Application application) {
      return DaggerAppComponent.builder()
        .appModule(new AppModule(application))
        .leakCanaryModule(new LeakCanaryModule(application))
        .build();
    }
  }
}

File: src/main/java/burrows/apps/example/gif/di/component/AppComponent.kt
package burrows.apps.example.gif.di.component

import android.app.Application
import burrows.apps.example.gif.App
import burrows.apps.example.gif.di.module.ActivityBuilderModule
import burrows.apps.example.gif.di.module.AppModule
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjector
import dagger.android.support.AndroidSupportInjectionModule
import javax.inject.Singleton

@Singleton
@Component(modules = [AndroidSupportInjectionModule::class, AppModule::class, ActivityBuilderModule::class])
interface AppComponent : AndroidInjector<App> {
  @Component.Builder
  interface Builder {
    @BindsInstance
    fun application(application: Application): AppComponent.Builder
    fun build(): AppComponent
  }
}
